File: 141-linked-list-cycle/141-linked-list-cycle.java
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public boolean hasCycle(ListNode head) {
        
        if(head==null||head.next==null) return false;
        
         ListNode sp=head;       
         ListNode fp=head;     
        
        do{
          sp=sp.next;
            fp=fp.next.next;
        }while(sp!=fp && fp!=null && fp.next!=null);
        return sp==fp;
    }
}

File: 141-linked-list-cycle/141-linked-list-cycle.py
class Solution:
    def hasCycle(self, head: Optional[ListNode]) -> bool:
        if not head or not head.next: return False
        fp=head.next.next
        sp=head.next
        while sp!=fp and fp and fp.next:
            sp=sp.next
            fp=fp.next.next       
        return fp==sp